Mathlib/Data/Pi/Projections.lean

```diff
@@ -0,0 +1,232 @@
+Authors: Etienne Marion
+-/
+import Mathlib.MeasureTheory.MeasurableSpace.Basic
+import Mathlib.Order.Interval.Finset.Nat
+import Mathlib.Topology.Constructions
+
+/-!
+# Projection of a function
+
+Given a dependent function `f : (i : ι) → α i` and `s : Set ι`, one might want to consider
+the restriction of `f` to the variables contained in `s`.
+This is defined in this file as `Function.proj s f`.
+Similarly, if we have `s t : Set ι`, `hst : s ⊆ t` and `f : (i : ↑t) → α ↑i`,
+one might want to restrict it to variables contained in `s`.
+This is defined in this file as `Function.proj₂ hst f`.
+We also define analoguous functions when `Finset`s are provided.
+
+Having these definitions avoids heavy type ascription when manipulating these functions and allow
+some rewriting when composing them (see `Function.proj₂_comp_proj` for instance).
+
+We also define analoguous functions for sets of the form `Set.Iic n` and `Finset.Iic n`, where
+`n : ℕ`, see `Function.projNat` for instance.
+
+## Main definitions
+
+* `Function.proj s f`: Restricts the function `f` to the variables contained in `s`.
+* `Function.projNat`: Restricts the function `f` to the variables indexed by integers
+  smaller than `n`.
+-/
+
+open MeasureTheory Function
+
+section Generality
+
+variable {ι : Type*}
+
+namespace Function
+
+variable {α : ι → Type*}
+
+/-- Given a dependent function, restrict it to a function of variables in `s`. -/
+@[simp]
+def proj (s : Set ι) (x : (i : ι) → α i) (i : s) : α i := x i
+
+/-- Given a dependent function of variables in `t`, restrict it to a function of variables in `s`
+when `s ⊆ t`. -/
+@[simp]
+def proj₂ {s t : Set ι} (hst : s ⊆ t) (x : (i : t) → α i) (i : s) : α i := x ⟨i.1, hst i.2⟩
+
+/-- Given a dependent function, restrict it to a function of variables in `s`, `Finset` version. -/
+@[simp]
+def fproj (s : Finset ι) (x : (i : ι) → α i) (i : s) : α i := x i
+
+/-- Given a dependent function of variables in `t`, restrict it to a function of variables in `s`
+when `s ⊆ t`, `Finset` version. -/
+@[simp]
+def fproj₂ {s t : Finset ι} (hst : s ⊆ t) (x : (i : t) → α i) (i : s) : α i := x ⟨i.1, hst i.2⟩
+
+theorem fproj_eq_proj (s : Finset ι) : @fproj ι α s = proj (s : Set ι) := rfl
+
+theorem frpoj₂_eq_proj₂ {s t : Finset ι} (hst : s ⊆ t) :
+    fproj₂ (α := α) hst = fproj₂ (Finset.coe_subset.2 hst) := rfl
+
+theorem proj₂_comp_proj {s t : Set ι} (hst : s ⊆ t) :
+    (proj₂ hst) ∘ (@proj ι α t) = proj s := rfl
+
+theorem fproj₂_comp_fproj {s t : Finset ι} (hst : s ⊆ t) :
+    (fproj₂ hst) ∘ (@fproj ι α t) = fproj s := rfl
+
+theorem proj₂_comp_proj₂ {s t u : Set ι} (hst : s ⊆ t) (htu : t ⊆ u) :
+    (proj₂ (α := α) hst) ∘ (proj₂ htu) = proj₂ (hst.trans htu) := rfl
+
+theorem fproj₂_comp_fproj₂ {s t u : Finset ι} (hst : s ⊆ t) (htu : t ⊆ u) :
+    (fproj₂ (α := α) hst) ∘ (fproj₂ htu) = fproj₂ (hst.trans htu) := rfl
+
+end Function
+
+variable {X : ι → Type*}
+
+section measurability
```

Comment:
Since the basic notation is very general, I would put it in a file with low imports, to make sure it can be used early on. And I would move the parts that need measurability to another file, probably in the `MeasureTheory` folder. And the parts about continuity to another file in the `Topology` folder.

Reply:
> Since the basic notation is very general, I would put it in a file with low imports

You mean I should move the measurability and continuity section so that the imports of the def are low right?

Reply:
I moved them close to similar results for `Function.update`.

Reply:
Hmmm this creates a conflict because of an `assert_not_exists Nat.instLocallyFiniteOrder` in `EMetricSpace.Defs`. Should I move continuity results to their own file then?

Reply:
Yes, if you can't find a suitable place, just create a new file. But I think you should also split the original file, into the very general part that doesn't depend on Nat, and the Nat-specific one which has much stronger imports.